### PR TITLE
partner: Add Upstage partner package

### PR DESCRIPTION
**Pull request by JuHyung-Son**
**Description**: add Upstage partner packages including UpstageChat, UpstageEmbeddings
**Dependencies**: openai
**Twitter handle**: [@upstageai](https://twitter.com/upstageai)

- [x]  **Add tests and docs**: If you're adding a new integration, please include
    1. a test for the integration, preferably unit tests that do not rely on network access,
    2. an example notebook showing its use. It lives in `docs/docs/integrations` directory.
- [x]  **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified. See contribution guidelines for more: https://python.langchain.com/docs/contributing/

Additional guidelines:

- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, hwchase17.

### CODE REVIEW

File: docs/docs/integrations/llms/solar.ipynb
Comment (efriis) [inline docs/docs/integrations/llms/solar.ipynb:1]: should we keep this? iirc there was also an LLM (text completion) integration in addition to chat model

File: docs/docs/integrations/text_embedding/solar.ipynb
Comment (efriis) [inline docs/docs/integrations/text_embedding/solar.ipynb:1]: need a redirect in vercel.json for the renamed ones (solar -> upstage)

File: docs/docs/integrations/tools/upstage_groundedness_check.ipynb
Comment (efriis) [inline docs/docs/integrations/tools/upstage_groundedness_check.ipynb:1]: do this in followup pr with this implementation (remove from this PR)

File: libs/community/langchain_community/chat_models/solar.py [hunk 1]
```diff
@@ -2,16 +2,19 @@
...
 from langchain_community.chat_models import ChatOpenAI
 from langchain_community.llms.solar import SOLAR_SERVICE_URL_BASE, SolarCommon
 
 
-class SolarChat(SolarCommon, ChatOpenAI):  # type: ignore[misc]
-    """Solar large language models.
-
+@deprecated(
+    since="0.0.32", removal="0.2.0", alternative_import="langchain_upstage.ChatUpstage"
+)
```
Comment (efriis) [inline libs/community/langchain_community/chat_models/solar.py:15]:
```suggestion
@deprecated(
    since="0.0.34", removal="0.2.0", alternative_import="langchain_upstage.ChatUpstage"
)
```

File: libs/community/langchain_community/embeddings/solar.py [hunk 2]
```diff
@@ -44,6 +45,9 @@ def _embed_with_retry(*args: Any, **kwargs: Any) -> Any:
     return _embed_with_retry(*args, **kwargs)
 
 
+@deprecated(
+    since="0.0.32", removal="0.2.0", alternative_import="langchain_upstage.ChatUpstage"
+)
```
Comment (efriis) [inline libs/community/langchain_community/embeddings/solar.py:50]:
```suggestion
@deprecated(
    since="0.0.34", removal="0.2.0", alternative_import="langchain_upstage.ChatUpstage"
)
```

File: libs/partners/upstage/LICENSE [hunk 3]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2023 LangChain, Inc.
```
Comment (efriis) [inline libs/partners/upstage/LICENSE:3]:
```suggestion
Copyright (c) 2024 LangChain, Inc.
```

File: libs/partners/upstage/langchain_upstage/chat_models.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    def _llm_type(self) -> str:
+        """Return type of chat model."""
+        return "upstage-chat"
+
+    model_name: str = Field(default="solar-1-mini-chat", alias="model")
+    """Model name to use."""
+    upstage_api_key: Optional[SecretStr] = Field(default=None, alias="api_key")
+    """Automatically inferred from env are `UPSTAGE_API_KEY` if not provided."""
+    upstage_api_base: Optional[str] = Field(
+        default="https://api.upstage.ai/v1/solar", alias="base_url"
```
Comment (efriis) [inline libs/partners/upstage/langchain_upstage/chat_models.py:60]: if this is encoded in the SDK as well, I would prefer to make this default to `None` (and let the sdk set the default)

File: libs/partners/upstage/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+repository = "https://github.com/langchain-ai/langchain"
+license = "MIT"
+
+[tool.poetry.urls]
+"Source Code" = "https://github.com/langchain-ai/langchain/tree/master/libs/partners/upstage"
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<4.0"
+openai = "^1.10.0"
+langchain-core = ">=0.0.12"
```
Comment (efriis) [inline libs/partners/upstage/pyproject.toml:16]:
```suggestion
langchain-core = "^0.1.42"
```

File: libs/partners/upstage/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+langchain-openai = ">=0.1.0"
```
Comment (efriis) [inline libs/partners/upstage/pyproject.toml:17]:
```suggestion
langchain-openai = "^0.1.3"
```

File: libs/partners/upstage/pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+freezegun = "^1.2.2"
+pytest-mock  = "^3.10.0"
+syrupy = "^4.0.2"
+pytest-watcher = "^0.3.4"
+pytest-asyncio = "^0.21.1"
+langchain-openai = {path = "../openai", develop = true}
+langchain-core = {path = "../../core", develop = true}
+docarray = "^0.32.1"
+pydantic = "^1.10.9"
+langchain_community = {path = "../../community", develop = true}
```
Comment (efriis) [inline libs/partners/upstage/pyproject.toml:33]: why is this necessary?

Comment (efriis) [reply libs/partners/upstage/pyproject.toml:33]: let's remove this and the rag integration test. `test_embeddings` is sufficient

File: libs/partners/upstage/tests/integration_tests/test_rag.py
Comment (efriis) [inline libs/partners/upstage/tests/integration_tests/test_rag.py:1]: remove - testing separate component interfaces is sufficient

File: libs/partners/upstage/tests/integration_tests/test_chat_models.py
Comment (efriis) [inline libs/partners/upstage/tests/integration_tests/test_chat_models.py:1]: Can we also add a `test_chat_models_standard.py` implementing the standard tests? See openai/anthropic for example of how to do that

Should be both the unit test stub as well as the integration test stub (they're new! Will be in bootstrapped package soon)

File: poetry.lock
Comment (efriis) [inline poetry.lock:1]: revert

File: pyproject.toml
Comment (efriis) [inline pyproject.toml:1]: revert

**Review by efriis — COMMENTED** [review]
great implementation! Some comments:

File: docs/docs/integrations/llms/solar.ipynb
Comment (JuHyung-Son) [reply docs/docs/integrations/llms/solar.ipynb:1]: While we are rebranding Solar to Upstage, we decided to eliminate the LLM integration and solely use the chat model. Do you think we need to integrate LLM?

And should we remove these documents after the Solar community package has been deprecated?

File: docs/docs/integrations/text_embedding/solar.ipynb
Comment (JuHyung-Son) [reply docs/docs/integrations/text_embedding/solar.ipynb:1]: Thank you. I have updated the vercel.json file; please review it.

File: libs/partners/upstage/tests/integration_tests/test_chat_models.py
Comment (JuHyung-Son) [reply libs/partners/upstage/tests/integration_tests/test_chat_models.py:1]: i added two tests

File: docs/docs/integrations/llms/solar.ipynb
Comment (chosh0615) [reply docs/docs/integrations/llms/solar.ipynb:1]: LLM integration is a legacy and we only support Chat model currently. We thought it'd be more appropriate to deprecate the community version of LLM integration and remove the notebook in this PR.

File: libs/partners/upstage/langchain_upstage/chat_models.py (hunk 4 above)
Comment (JuHyung-Son) [reply libs/partners/upstage/langchain_upstage/chat_models.py:60]: Good point! We use the OpenAI SDK, so its default value is OpenAI's API endpoint. That is why we set the default value here.

File: docs/docs/integrations/llms/solar.ipynb
Comment (chosh0615) [reply docs/docs/integrations/llms/solar.ipynb:1]: Like we chatted, we've put this file back with a note of deprecation!

File: docs/vercel.json
Comment (efriis) [inline docs/vercel.json:1]: need a redirect for chat too

File: libs/partners/upstage/tests/unit_tests/test_chat_models_standard.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+class TestUpstageStandard(ChatModelUnitTests):
+    @pytest.fixture
+    def chat_model_class(self) -> Type[BaseChatModel]:
+        return ChatUpstage
+
+    @pytest.fixture
+    def chat_model_params(self) -> dict:
+        return {
+            "model": "solar-1-mini-embedding",
```
Comment (efriis) [inline libs/partners/upstage/tests/unit_tests/test_chat_models_standard.py:19]: should probably be the chat model name

**Review by efriis — COMMENTED** [review]
Also needs a `poetry lock` in the partner package looks like.

Will take this over the finish line in a new PR! In the future, if you open from a personal branch instead of an organization branch, we can help edit your PR in place :)

https://python.langchain.com/docs/contributing/faq/#how-do-i-allow-maintainers-to-edit-my-pr

File: docs/vercel.json
Comment (JuHyung-Son) [reply docs/vercel.json:1]: I also need to redirect the LLMS, correct?

File: libs/partners/upstage/tests/unit_tests/test_chat_models_standard.py (hunk 8 above)
Comment (JuHyung-Son) [reply libs/partners/upstage/tests/unit_tests/test_chat_models_standard.py:19]: Nice catch. Thank you!

**Comment by efriis** [comment]
Working in #20574 (see review comment for link to how to collaborate in future on a personal branch!)
